### DISCUSSION TITLE
Spurious PyPy crashes in nanobind CI test matrix

### DISCUSSION THREAD

**Discussion by wjakob** [pypy/pypy#4923] [Q&A]
Dear PyPy team,

I maintain the [nanobind](https://github.com/wjakob/nanobind) C++ <-> Python interop layer that also ensures PyPy cpyext compatibility through some entries in its build matrix. For a long time now, we've had issues with PyPy crashing spuriously while loading the extensions used in its test suite. You can see an example of such a crash here: https://github.com/wjakob/nanobind/actions/runs/8234561874/job/22516568891

It's unfortunately not very much to go by. Worse, I haven't been able to replicate this crash with separate testing on an intel Linux machine running exactly the same version of PyPy. Hence I wanted to ask a few questions:

1. Are there any known stability issues with PyPy or cpyext that could explain this?
2. Have you encountered such crashes on GitHub actions? (using the `setup-python` action). I am wondering if there is a difference between the binary they provide and what I download from the PyPy webpage.
3. Did you ever have such an issue with crashes only happening on GitHub, and how did you debug them? (I know that @mattip has a separate testing repository for such purposes)

My hope with this ticket was to start a discussion and figure out what's wrong.

Thanks,
Wenzel

**Comment by mattip** [comment]
This looks like a failure in test collection. Maybe @cfbolz or @rlamy have seen some of this. It might be some problem with pytest. Did you see this with earlier versions of pytest?

**Reply by wjakob** [reply]
PyTest is written in pure Python code and so shouldn't be able to crash the interpreter regardless of the implementation. PyTest is importing nanobind-based extensions here.

In contrast to normal C Python extensions that are relatively static (pointer to a table of methods/classes), nanobind is more dynamic. It dynamically creates a custom Python object for every function binding, a custom metaclass, and then it uses `PyType_FromMetaclass` to create the class bindings.

I suspect that it's this sequence that causes the crash, either because of nanobind making assumptions that are too specific to CPython, or a bug in PyPy's implementation of the CPython API. But I don't have a good handle on how to debug this since it only happens in the cloud 😓. Did you notice such crashes as well @mattip ? If I remember correctly, you are running the nanobind test suite against PyPy HEAD.

**Comment by mattip** [comment]
The [weekly testing runs](https://github.com/pypy/binary-testing/actions/workflows/nanobind.yml) look pretty clean except for one week where `typing_extensions` was added as a dependency. Are the runs equivalent to your CI?

**Reply by wjakob** [reply]
You're missing the NumPy and Eigen linear algebra library tests (skipped, likely because of missing dependencies). However, that is fine I think. The "contact surface" between nanobind and PyPy is well-covered by the other test.

**Comment by cfbolz** [comment]
Hi Wenzel! From the traceback it looks like it's another instance of our 'pytest assert rewriting mystery bug': https://github.com/pypy/pypy/issues/3959
And more recently: https://github.com/PyO3/pyo3/issues/3766

Therefore this very likely has nothing to do with nanobind directly. 

I'm incredibly frustrated with this issue, so far none of the instances have been reproducible on anyones laptop, they happen only on CI. therefore I haven't made much progress on understanding what's going wrong, despite various attempts :-(.

Something that I've been meaning to try is to change one of the involved CI scripts to run this PyPy build: https://buildbot.pypy.org/pypy/pypy-lldebug-nightly.tar.bz2
(it's a 3.10 Linux build iirc). It has a bunch of extra internal assertions turned on that might potentially help track the problem down somewhat more. How reliably does the nanobind test fail? Maybe I should try to fork nanobind and whack at the CI script a bit to see what happens.

Another option that I've been thinking about is to get CI to produce a coredump somehow and upload that as an artifact.

**Reply by wjakob** [reply]
Dear @cfbolz,

oh wow, I did not realize this was such a big thing. We run 6 PyPy configurations (two versions of PyPy -- each on macOS, Linux, and Windows). The build matrix fails with P=~1/3. Most often it's the linux build that crashes in my experience. Probably you could simply clone our repository with the GHA build matrix and replicate the linux configuration 10 times to have a solid repro.

**Comment by mattip** [comment]
I see the nanobind CI run installs Eigen, but not NumPy, so I will add that to the binary-testing CI run. The binary-testing run has an extra "Setup annotations on Linux" step, should that be removed?

Looking back at the failures in [the workflow](https://github.com/wjakob/nanobind/actions/workflows/ci.yml), I see:
time | ci fail | error message
---|---|---
March 12, 2024, 09:18| all runs | [`ImportError: Overload was leaked!`](https://github.com/wjakob/nanobind/actions/runs/8244986357/job/22548132378#step:9:117)
. | no | 4 successful runs
March 11 2024, 17:30 | CPython error
. | no | 1 successful run
MArch 11 2024, 16:30 | [Python pypy3.10-v7.3.15 / ubuntu-latest](https://github.com/wjakob/nanobind/actions/runs/8234561874/job/22516568891#step:10:58) | segfault
. | no | 6 successful runs
March 10 2024, 12:47 | all, plus CPython3.13 | missing NumPy
. | no | 3 successful runs
March 10 2024, 07:43 | [Python pypy3.9-v7.3.12 / ubuntu-latest](https://github.com/wjakob/nanobind/actions/runs/8219889254/job/22478370496#logs) | segfault
. | no | 1 successful run
March 10 2024 7:25 | [Python pypy3.9-v7.3.12 / windows-2022](https://github.com/wjakob/nanobind/actions/runs/8219802329/job/22478195101#logs) | segfault? (no logs)
March 10 2024 7:17 | [Python pypy3.9-v7.3.12 / ubuntu-latest](https://github.com/wjakob/nanobind/actions/runs/8219754832/job/22478111705#logs) | segfault

Going back further, I see more segfaults, mostly on ubuntu, I couldn't find any on macos. I wonder why the binary-testing CI runs don't see any failures when here they occur pretty often.

Maybe it is something else entirely like the setup-cmake version?

**Comment by mattip** [comment]
The first CI run of pypy/binary-testing#49, which resyncs the CI workflow with upstream, crashed. I commented there on two things that might have changes: using a different cmake github action and some different build macros. Thoughts?

**Reply by wjakob** [reply]
Great. that's exactly the kind of crash that we are talking about here (https://github.com/pypy/binary-testing/actions/runs/8245574414/job/22549834434).

I think you can remove all the CPython entries from the matrix since they are not relevant. In turn, you could insert some "dummy dimensions" into the matrix to get it to test PyPy many times with the same configuration to have more shots at triggering the issue.

In the nanobind repo, I usually push the button to re-run crashed PyPy test runs when I get an email, and this re-run usually succeeds and causes things to be green. So the timeline on GitHub actions is not  representative of how often this fails on the nanobind repo.

I answered a question about compilation flags in the binary-testing PR.

**Comment by cfbolz** [comment]
just a quick update, I've managed to eventually reproduce it locally. the bug is still totally baffling, but at least it's not only on github ci any more. will write an issue to track it and document details how to reproduce it soon.

**Reply by cfbolz** [reply]
here's the issue: https://github.com/pypy/pypy/issues/4925
